### PR TITLE
New link protocol

### PR DESCRIPTION
**Pull request by rickard-green**
Introduce a new link protocol which prevents links from ending up in an inconsistent state where one participant considers itself linked while the other doesn't. This bug has always existed in the distributed case, but has since OTP 21 also existed in the node local case since the distributed link protocol then was adopted also for node local links.

Besides the introduction of the new link protocol, the documentation regarding links and exit signals has also been improved.

* The [new link protocol](https://erlang.org/~rickard/OTP-17127/erts-11.1.8/doc/html/erl_dist_protocol.html#new_link_protocol)
* The [`link/1`](https://erlang.org/~rickard/OTP-17127/erts-11.1.8/doc/html/erlang.html#link-1) BIF
* The [`unlink/1`](https://erlang.org/~rickard/OTP-17127/erts-11.1.8/doc/html/erlang.html#unlink-1) BIF
* [Links](https://erlang.org/~rickard/OTP-17127/doc/reference_manual/processes.html#links)
* [Sending exit signals](https://erlang.org/~rickard/OTP-17127/doc/reference_manual/processes.html#sending_exit_signals)
* [Receiving exit signals](https://erlang.org/~rickard/OTP-17127/doc/reference_manual/processes.html#receiving_exit_signals)

---
Code merged into `maint` branch 10/3-2021

### CODE REVIEW

File: erts/doc/src/erl_dist_protocol.xml [hunk 1]
```diff
@@ -1096,6 +1096,15 @@ DiB == gen_digest(ChA, ICA)?
           <p>The node supports the new connection setup handshake (version 6)
 	  introduced in OTP 23.</p>
         </item>
+        <tag><marker id="DFLAG_UNLINK_ID"/><c>-define(DFLAG_UNLINK_ID, 16#2000000).</c></tag>
+        <item>
+	  <p>Use the <seeguide marker="#new_link_protocol">new link protocol</seeguide>.</p>
+	  <note><p>This flag will become mandatory in OTP 26.</p></note>
+	  <p>When this flag is set, the link protocol previously used is
+	  replaced by a new link protocol. This since the previously used
+	  protocol could cause process links to end up in an inconsistent
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1105]:
```suggestion
	  <p>When this flag is set, the previously used link protocol is
	  replaced by a new link protocol. The former
	  could cause process links to end up in an inconsistent
```

Comment (paulo-ferraz-oliveira) [reply erts/doc/src/erl_dist_protocol.xml:1105]: I still don't think it properly conveys the change; I just tried to clear it up.

Comment (rickard-green) [reply erts/doc/src/erl_dist_protocol.xml]: I've added an explanation of what the distribution flags/capability flags are in the beginning of that section and changed the documentation of `DFLAG_UNLINK_ID`. Hope that clears it up.

File: erts/doc/src/erl_dist_protocol.xml [hunk 2]
```diff
@@ -974,7 +974,11 @@ DiB == gen_digest(ChA, ICA)?
     <section>
       <marker id="dflags"/>
       <title>Distribution Flags</title>
-      <p>The following capability flags are defined:</p>
+      <p>Early in the distribution handshake the two participating nodes
+      exchange capability flags. This in order to determine how the
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:978]:
```suggestion
      exchange capability flags. This is done in order to determine how the
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 3]
```diff
@@ -1096,6 +1100,14 @@ DiB == gen_digest(ChA, ICA)?
           <p>The node supports the new connection setup handshake (version 6)
 	  introduced in OTP 23.</p>
         </item>
+        <tag><marker id="DFLAG_UNLINK_ID"/><c>-define(DFLAG_UNLINK_ID, 16#2000000).</c></tag>
+        <item>
+	  <p>Use the <seeguide marker="#new_link_protocol">new link protocol</seeguide>.</p>
+	  <note><p>This flag will become mandatory in OTP 26.</p></note>
+	  <p>Unless both nodes has set the <c>DFLAG_UNLINK_ID</c> flag, the
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1107]:
```suggestion
	  <p>Unless both nodes have set the <c>DFLAG_UNLINK_ID</c> flag, the
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 4]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	<seeguide marker="#UNLINK"><c>UNLINK</c></seeguide>
+	signal. The old <seeguide marker="#LINK"><c>LINK</c></seeguide>
+	signal is still sent in order to set up a link, but handled
+	differently upon reception.
+      </p>
+
+      <p>
+	In order to set up a link, a <c>LINK</c> signal is sent, from
+	the process initiating the operation, to the process that it
+	wants to link with. In order to remove a link, an
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1614]:
```suggestion
	wants to link to. In order to remove a link, an
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 5]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	<c>UNLINK_ID</c> signal is sent, from the process initiating
+	the operation, to the linked process. The receiver of an
+	<c>UNLINK_ID</c> signal responds with an <c>UNLINK_ID_ACK</c>
+	signal. Upon reception of an <c>UNLINK_ID</c> signal, the
+	corresponding <c>UNLINK_ID_ACK</c> signal <em>must</em> be
+	sent before any other signals are sent to the sender of the
+	<c>UNLINK_ID</c> signal. This together with
+	<seeguide marker="erts:communication#passing-of-signals">the
+	signal ordering guarantees</seeguide> of Erlang makes it
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1623]:
```suggestion
	<c>UNLINK_ID</c> signal. Together with
	<seeguide marker="erts:communication#passing-of-signals">the
	signal ordering guarantees</seeguide> of Erlang this makes it
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 6]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	possible for the sender of the <c>UNLINK_ID</c> signal to know
+	the order of other signals which is essential for the protocol.
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1625]:
```suggestion
	the order of other signals which are essential for the protocol.
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 7]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	The <c>UNLINK_ID_ACK</c> signal should contain the same
+	<c>Id</c> as the <c>Id</c> contained in the <c>UNLINK_ID</c>
+	signal being acknowledged.
+      </p>
+
+      <p>
+	Processes also need to maintain process local information about
+	links. The state of this process local information is changed
+	when the signals above are sent and received. This process
+	local information also determine if a signal should be sent
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1635]:
```suggestion
	local information also determines if a signal should be sent
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 8]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	when a process calls
+	<seemfa marker="erlang#link/1"><c>link/1</c></seemfa> or
+	<seemfa marker="erlang#unlink/1"><c>unlink/1</c></seemfa>.
+	A <c>LINK</c> signal is only sent if there currently does not
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1639]:
```suggestion
	A <c>LINK</c> signal is only sent if there does not currently
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 9]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	<item>
+	  Process identifier of the linked process.
+	</item>
+	<tag>Active Flag</tag>
+	<item>
+	  If set, the link is active and the process will react on
+	  <seeguide marker="system/reference_manual:processes#receiving_exit_signals">incoming
+	  exit signals</seeguide> issued due to the link. If not set,
+	  the link is inactive and incoming exit signals issued due
+	  to the link will be ignored. That is, the processes are
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1660]:
```suggestion
	  the link is inactive and incoming exit signals, issued due
	  to the link, will be ignored. That is, the processes are
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 10]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	  considered as <em>not</em> linked.
+	</item>
+	<tag>Unlink Id</tag>
+	<item>
+	  Identifier of an outstanding unlink operation. That is,
+	  an unlink operation that has not been acknowledged yet.
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1666]:
```suggestion
	  an unlink operation that has not yet been acknowledged.
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 11]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	  active flag is set, and unlink id is cleared. That is,
+	  if we had an outstanding unlink operation we will ignore
+	  the result of that operation and enable the link.
+	</item>
+	<tag>A <c>LINK</c> signal is received</tag>
+	<item>
+	  If no link information already exists, it is created, the
+	  active flag is set and unlink id is cleared. If the link
+	  information already exists, the signal is silently ignored.
+	  This regardless of whether the active flag is set or not.
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1696]:
```suggestion
	  information already exists, the signal is silently ignored,
	  regardless of whether the active flag is set or not.
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 12]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	  That is, if we have an outstanding unlink operation we will
+	  <em>not</em> activate the link. In this scenario, the sender
+	  of the <c>LINK</c> signal has not yet sent an
+	  <c>UNLINK_ID_ACK</c> signal corresponding to our
+	  <c>UNLINK_ID</c> signal which means that it will receive
+	  the our <c>UNLINK_ID</c> signal after it sent this
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1702]:
```suggestion
	  our <c>UNLINK_ID</c> signal after it sent its
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 13]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	<item>
+	  If the active flag is set, information about the link
+	  is removed. If the active flag is not set (that is, we have
+	  an outstanding unlink operation), the information about the
+	  link is left unchanged.
+	</item>
+	<tag>An <c>UNLINK_ID_ACK</c> signal is sent</tag>
+	<item>
+	  This is done when an <c>UNLINK_ID</c> signal is received and
+	  cause no further changes of the link information.
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1723]:
```suggestion
	  causes no further changes of the link information.
```

File: erts/doc/src/erl_dist_protocol.xml [hunk 14]
```diff
@@ -1513,6 +1536,247 @@ DiB == gen_digest(ChA, ICA)?
...
+	</item>
+	<tag>An <c>UNLINK_ID_ACK</c> signal is received</tag>
+	<item>
+	  If information about the link exist, the active flag is not
```
Comment (paulo-ferraz-oliveira) [inline erts/doc/src/erl_dist_protocol.xml:1727]:
```suggestion
	  If information about the link exists, the active flag is not
```

File: system/doc/reference_manual/processes.xml [hunk 15]
```diff
@@ -122,14 +122,39 @@ spawn(Module, Name, Args) -> pid()
...
-      between two processes <c>Pid1</c> and <c>Pid2</c> is created
-      by <c>Pid1</c> calling the BIF <c>link(Pid2)</c> (or conversely).
-      There also exist a number of <c>spawn_link</c> BIFs, which spawn
-      and link to a process in one operation.</p>
-    <p>Links are bidirectional and there can only be one link between
-      two processes. Repeated calls to <c>link(Pid)</c> have no effect.</p>
-    <p>A link can be removed by calling the BIF <c>unlink(Pid)</c>.</p>
+    <p>
+      Two processes can be <em>linked</em> to each other. Also a
+      process and a port that resides on the same node can be linked
```
Comment (paulo-ferraz-oliveira) [inline system/doc/reference_manual/processes.xml:127]:
```suggestion
      process and a port that reside on the same node can be linked
```

File: system/doc/reference_manual/processes.xml [hunk 16]
```diff
@@ -122,14 +122,39 @@ spawn(Module, Name, Args) -> pid()
...
+      to each other. A link beteen two processes can be created
+      if one of them calls the
+      <seemfa marker="erts:erlang#link/1"><c>link/1</c></seemfa> BIF
+      with the process identifier of the other process as argument.
+      Links can also be created using one the following spawn BIFs
+      <seemfa marker="erts:erlang#spawn_link/4"><c>spawn_link()</c></seemfa>,
+      <seemfa marker="erts:erlang#spawn_opt/5"><c>spawn_opt()</c></seemfa>, or
+      <seemfa marker="erts:erlang#spawn_request/5"><c>spawn_request()</c></seemfa>.
+      The spawn operation and the link operation will in these cases
+      be performed atomically.
```
Comment (paulo-ferraz-oliveira) [inline system/doc/reference_manual/processes.xml:137]:
```suggestion
      The spawn operation and the link operation will 
      be performed atomically, in these cases.
```

File: erts/doc/src/erl_dist_protocol.xml (hunk 6 above)
Comment (rickard-green) [reply erts/doc/src/erl_dist_protocol.xml:1625]: Hm, I think it should be "is" here. It refers to "the order" not "signals".

**Comment by rickard-green** [comment]
Thanks @paulo-ferraz-oliveira for the language fixes! I included most of them.

**Pull request merged**